code/datums/components/team_monitor.dm

```diff
@@ -0,0 +1,535 @@
+	if(!frequency_added)
+		return
+	if(islist(GLOB.tracker_huds[frequency_added]))
+		GLOB.tracker_huds[frequency_added] |= component_added
+	else
+		GLOB.tracker_huds[frequency_added] = list(component_added)
+
+//Adds a new tracking beacon
+/proc/add_tracker_beacon(frequency_added, datum/component/component_added)
+	if(!frequency_added)
+		return
+	if(islist(GLOB.tracker_beacons[frequency_added]))
+		GLOB.tracker_beacons[frequency_added] |= component_added
+	else
+		GLOB.tracker_beacons[frequency_added] = list(component_added)
+
+/proc/get_all_beacons_on_frequency(frequency, base_frequency)
+	if(!frequency)
+		return list()
+	var/list/found_beacons = list()
+	if(islist(GLOB.tracker_beacons[frequency]))
+		found_beacons.Add(GLOB.tracker_beacons[frequency])
+	if(islist(GLOB.tracker_beacons["[base_frequency]-GLOB"]))
+		found_beacons.Add(GLOB.tracker_beacons["[base_frequency]-GLOB"])
+	return found_beacons
+
+/proc/get_all_watchers_on_frequency(frequency, team_key = "", global_freq = FALSE)
+	if(global_freq)
+		. = list()
+		for(var/tracker_freq in GLOB.tracker_huds)
+			for(var/datum/component/team_monitor/TM as() in GLOB.tracker_huds[tracker_freq])
+				if(TM.team_freq_key == team_key)
+					. += TM
+	else
+		return GLOB.tracker_huds[frequency]
+
+//==================
+// Component
+//  - HUD COMPONENT
+//  - HANDLES POINTING TOWARDS TRACKED BEACONS
+//==================
+
+//The component that handles tracking atoms
+/datum/component/team_monitor
+	//The frequency of the team signals we are trackings
+	var/team_freq_key = "debug"	//Key <-- cannot be changed
+	var/team_frequency = ""	//Final compiled: Consists of key then numbers between 1 and 999
+	//The atoms we are actually tracking
+	// Key = Beacon component
+	// Value = image
+	var/list/tracking = list()
+	//Who are we updating for
+	var/mob/updating = null
+	//Distance from center
+	//Probably in pixels or something idk
+	var/distance = 20
+	//Should we display the hud in the firstplace
+	var/hud_visible = TRUE
+	//The attached beacon: Ignore this one
+	var/datum/component/tracking_beacon/attached_beacon
+
+/datum/component/team_monitor/Initialize(frequency_key, frequency, _attached_beacon)
+	var/obj/item/clothing/item = parent
+	if(!istype(item))
+		return COMPONENT_INCOMPATIBLE
+
+	team_freq_key = frequency_key
+	if(frequency != null)
+		team_frequency = "[frequency_key][frequency]"
+	else
+		team_frequency = null
+
+	attached_beacon = _attached_beacon
+
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/parent_equipped)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/parent_dequpped)
+
+	get_matching_beacons()
+	add_tracker_hud(team_frequency, src)
+
+/datum/component/team_monitor/Destroy(force, silent)
+	//Unregister signals
+	if(parent)
+		UnregisterSignal(parent, COMSIG_ITEM_EQUIPPED)
+		UnregisterSignal(parent, COMSIG_ITEM_DROPPED)
+
+	if(team_frequency)
+		GLOB.tracker_huds[team_frequency] -= src
+
+	//Stop processing
+	STOP_PROCESSING(SSprocessing, src)
```

Comment:
I suggest a more thorough cleaning. You also have to clean `attached_beacon`.

```suggestion
	if(updating)
		hide_hud(updating)
	if(attached_beacon)
		if(attached_beacon.attached_monitor == src)
			attached_beacon.attached_monitor = null
		attached_beacon = null
```

Reply:
Beacons and HUDs aren't necessarily destroyed together

Reply:
ill have a look at better cleaning

Reply:
Ah yea attached_beacon needs to be set to null